netaddr.go

```diff
@@ -780,6 +788,43 @@ func (ip IP) string6() string {
 	return string(ret)
 }
 
+// StringExpanded returns the string form of the IP address ip, but IPv6
+// addresses are expanded with leading zeroes and no "::" compression. For
+// example, "2001:db8::1" becomes "2001:0db8:0000:0000:0000:0000:0000:0001".
+//
+// IPv4 and invalid IP addresses are formatted the same as when using IP.String.
+// See the documentation of IP.String for more details.
+func (ip IP) StringExpanded() string {
+	// Invalid and IPv4 are handled the same as in String.
+	switch ip.z {
+	case z0, z4:
+		return ip.String()
+	}
+
+	// Allocate enough space for the longest possible IPv6 address plus the
+	// exact length of the zone, if applicable.
+	max := len("ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff")
+	if ip.z != z6noz {
+		// Add space for % too.
+		max += 1 + len(ip.Zone())
+	}
+
+	ret := make([]byte, 0, max)
+	for i := uint8(0); i < 8; i++ {
```

Comment:
It feels tempting to unroll this loop for clarity, so that as a reader I don't have to parse out the `i > 0` special case and work things out. Probably not worth it though unless it leads to a massive inlining benefit or somesuch.

(i.e. no action required here, just thinking aloud)

Reply:
Will leave alone for now!

Reply:
Another option is to do appendHexPad then append : always, and then slice off the trailing : at the end. Not sure whether that is much clearer (or faster).